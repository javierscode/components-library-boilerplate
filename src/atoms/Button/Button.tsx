import React from 'react'
import classnames from 'classnames'
import styles from './Button.module.css'

export interface ButtonProps {
  /**
   * Is this the principal call to action on the page?
   */
  primary?: boolean
  /**
   * What background color to use
   */
  backgroundColor?: string
  /**
   * How large should the button be?
   */
  size?: 'small' | 'medium' | 'large'
  /**
   * Button contents
   */
  label: string
  /**
   * Optional click handler
   */
  onClick?: () => void
}

type StyleSize = 'button--small' | 'button--medium' | 'button--large'

/**
 * Primary UI component for user interaction
 */
const Button: React.FC<ButtonProps> = ({
  primary = false,
  size = 'medium',
  backgroundColor,
  label,
  ...props
}: ButtonProps) => {
  const StyleSize: StyleSize = `button--${size}`

  return (
    <button
      type='button'
      className={classnames(
        styles.button,
        styles[StyleSize],
        primary ? styles['button--primary'] : styles['button--secondary']
      )}
      style={{ backgroundColor }}
      {...props}
    >
      {label}
    </button>
  )
}

export default Button
