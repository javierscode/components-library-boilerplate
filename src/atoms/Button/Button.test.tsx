import React from 'react'
import Button from './Button'
import { cleanup, fireEvent, render } from '@testing-library/react'

afterEach(() => {
  cleanup()
})

describe('When the Button is rendered', () => {
  describe('And is Primary', () => {
    it('should match with the snapshot', () => {
      const { container } = render(<Button primary={true} label={'Snapshot'} />)
      expect(container).toMatchSnapshot()
    })
  })

  describe('And is Secondary', () => {
    it('should match with the snapshot', () => {
      const { container } = render(<Button primary={true} label={'Snapshot'} />)
      expect(container).toMatchSnapshot()
    })
  })
})

describe('When the Button is clicked', () => {
  it('should dispach the onclick function', () => {
    const mockedOnClick = jest.fn()

    const { getByRole } = render(
      <Button primary={true} label={'Snapshot'} onClick={mockedOnClick} />
    )

    fireEvent.click(getByRole('button'))

    expect(mockedOnClick).toBeCalledTimes(1)
  })
})
