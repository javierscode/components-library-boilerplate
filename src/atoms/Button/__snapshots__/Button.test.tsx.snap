// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When the Button is rendered And is Primary should match with the snapshot 1`] = `
<div>
  <button
    class="button button--medium button--primary"
    type="button"
  >
    Snapshot
  </button>
</div>
`;

exports[`When the Button is rendered And is Secondary should match with the snapshot 1`] = `
<div>
  <button
    class="button button--medium button--primary"
    type="button"
  >
    Snapshot
  </button>
</div>
`;
